File: Leetcode/Python/_98.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def traverse_and_check(node, min_val, max_val):
            if node is None:
                return True
            if node.val <= min_val or max_val <= node.val:
                return False
            return traverse_and_check(node.left, min_val, node.val) and traverse_and_check(node.right, node.val, max_val)

        
        return traverse_and_check(root, float("-Inf"), float("Inf"))

File: Leetcode/cpp/98.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        if (root == nullptr || (root->left == nullptr && root->right == nullptr)) {
            return true;
        }
        std::function<bool(TreeNode*, long long, long long)> dfs = [&](TreeNode* node, long long left, long long right) -> bool {
            if (node == nullptr) {
                return true;
            }
            if (left >= node->val || node->val >= right) {
                return false;
            }
            return dfs(node->left, left, node->val) && dfs(node->right, node->val, right);
        };
        return dfs(root, LLONG_MIN, LLONG_MAX);
    }
};
